### PR TITLE
New handling for `define:vars` scripts and styles

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Fixes `define:vars` handling for scripts and styles
- Requires https://github.com/withastro/compiler/pull/470
- For scripts with `define:vars`, we now wrap them in a unique scope with `{}` to avoid redeclaration errors
  - A previous version of this PR implemented new hoisting logic for `script` but this was deemed out of scope. We may revisit this in the future.
- For styles with `define:vars`, we can now ignore them entirely. They are handled at the compiler level.

## Testing

Tests updated

## Docs

Bug fix only

### CODE REVIEW

File: packages/astro/src/runtime/client/deserialize.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+interface PropTypeSelector {
```
Comment (matthewp) [inline packages/astro/src/runtime/client/deserialize.ts:1]: I don't see that this file is being used anywhere. What's it for?

Comment (natemoo-re) [reply packages/astro/src/runtime/client/deserialize.ts:1]: Whoops, leftover from some earlier work.

File: packages/astro/src/runtime/server/index.ts [hunk 2]
```diff
@@ -633,19 +619,31 @@ export function spreadAttributes(
...
+// converts (most) arbitrary strings to valid JS identifiers
+const toIdent = (k: string) =>
+  k.trim().replace(/(?:(?<!^)\b\w|\s+|[^\w]+)/g, (match, index) => {
+    if (/[^\w]|\s/.test(match)) return '';
+    return index === 0 ? match : match.toUpperCase();
+  });
+
 // Adds variables to an inline script.
-export function defineScriptVars(vars: Record<any, any>) {
+export async function defineScriptVars(vars: Record<any, any>) {
```
Comment (tony-sull) [inline packages/astro/src/runtime/server/index.ts:643]: does this need to be async?

**Review by tony-sull — APPROVED** [review]
lgtm!  One comment that shouldn't block a merge either way

**Pull request merged**